### PR TITLE
Fix Irrecoverable State for Queued to Sent Batch Failure Condition

### PR DESCRIPTION
**Pull request by willmeister**
## Description
Shortens window during which a process restart would require manual intervention to fix rollup batches.

## Questions
- Is there an easy way to get the tx hash ahead of sending so that we can eliminate this race condition? Note: my main problem is with estimating and setting `gasLimit` and `gasPrice` -- I know it's doable by doing a lot more of what ethers wraps manually, but this seems like the right ratio of improvement : time. 
- Does it make sense to just add wallet nonce to the batch submission table and update that ahead of sending a batch? That way, we can always check to see that the pending tx count is as expected. But if a nonce is stored, can't find the tx associated with it so 🤷 

### Fixes
- Fixes # [YAS-647](https://optimists.atlassian.net/browse/YAS-647)

## Contributing Agreement

- [x] I have read and understood the [Optimism Contributing Guide and Code of Conduct](https://github.com/ethereum-optimism/optimism-monorepo/blob/master/.github/CONTRIBUTING.md) and am following those guidelines in this pull request.

### CODE REVIEW

**Comment by karlfloersch** [comment]
You should be able to get the tx hash by hashing the output of: https://github.com/ethers-io/ethers.js/issues/535#issuecomment-498401610

**Comment by tynes** [comment]
`BatchSubmissionStatus.QUEUED` is when its in the mempool and `BatchSubmissionStatus.SUBMITTING` is when its being sent to the mempool but before its confirmed to enter?

**Comment by willmeister** [comment]
> `BatchSubmissionStatus.QUEUED` is when its in the mempool and `BatchSubmissionStatus.SUBMITTING` is when its being sent to the mempool but before its confirmed to enter?

* `BatchSubmissionStatus.QUEUED` it is guaranteed to have never been submitted to L1
* `BatchSubmissionStatus.SUBMITTING` we are trying to submit it, but if our process fails here, it may or may not have been submitted
* `BatchSubmissionStatus.SENT` sent and got at least 1 confirm on chain
* `BatchSubmissionStatus.FINALIZED` There were at least <configurable finality blocks> confirms

File: packages/rollup-core/src/app/data/consumers/canonical-chain-batch-submitter.ts [hunk 1]
```diff
@@ -78,39 +79,88 @@ export class CanonicalChainBatchSubmitter extends ScheduledTask {
...
+      this.canonicalTransactionChain,
+      'appendSequencerBatch',
+      [txsCalldata, timestamp, batchBlockNumber, startIndex],
+      this.submitterWallet
+    )
+  }
+
+  protected async getBatchSubmissionBlockNumber(): Promise<number> {
+    // TODO: This will eventually be part of the output metadata from L2 tx outputs
+    //  Need to update geth to have this functionality so this is a mock for now
```
Comment (karlfloersch) [inline packages/rollup-core/src/app/data/consumers/canonical-chain-batch-submitter.ts:160]: 👍 ah yes we definitely need this. Got to double check that we have it represented as a ticket

File: packages/rollup-core/src/app/data/consumers/state-commitment-chain-batch-submitter.ts [hunk 2]
```diff
@@ -81,17 +139,38 @@ export class StateCommitmentChainBatchSubmitter extends ScheduledTask {
...
+        `Appending state root batch number: ${stateRootBatch.batchNumber} with ${stateRoots.length} state roots at index ${stateRootBatch.startIndex}.`
       )
 
-      const txRes: TransactionResponse = await this.stateCommitmentChain.appendStateBatch(
+      const signedTx: string = await this.getSignedRollupBatchTx(
         stateRoots,
         stateRootBatch.startIndex
       )
+
+      txHash = keccak256(signedTx)
```
Comment (karlfloersch) [inline packages/rollup-core/src/app/data/consumers/state-commitment-chain-batch-submitter.ts:150]: Nice!

**Review by karlfloersch — APPROVED** [review]
Alright great! Seems it wasn't too verbose to calculate the tx hash up front & so that's good. And with the submission logic everything seems right to me!

LGTM!

**Pull request merged**